### DISCUSSION TITLE
elf and multiboot

### DISCUSSION THREAD

**Discussion by steve6375** [a1ive/grub2-filemanager#337] [General]
Sorry, but I don't understand this 'multiboot' concept and why grub_platform is set to multiboot - could you explain please?

Are you saying that if we boot to grub in Legacy mode on Virtual Box and grub_platform is set to "multiboot" then it is possible to also UEFI32 and UEFI64 boot to EFI files too, even though we are in legacy mode (and the CPU will be in real mode)? If so, how do we know which UEFI is supported UEFI64 or UEFI32?

Also, how to tell what mode we have booted in?
If boot using elf iso - e.g. if we want to boot to Windows can we boot in UEFI and Legacy mode?

P.S. The files in grubfm \samples test for $grub_platform == "pc"  - so they won't work if boot using elf iso?

**Comment by a1ive** [comment]
The value of the variable grub-platform is determined at compile time and it only indicates the platform of grub2 itself: pc, efi, uboot, multiboot, coreboot, xen ...
You can multiboot to grubfm.elf in both Legacy mode and UEFI ia32/x64, but grubfm.elf only has legacy bios disk driver (int13h), so it won't detect any disk in UEFI (no CSM).
If UEFI has int13h interrupt calls (CSM), grubfm.elf will load bios disk driver and detect disks.

grubfm will set variable `grub_mb_firmware` to `bios` if firmware has int13h, otherwise will set to `unknown`.
example: https://github.com/a1ive/grub2-filemanager/blob/cba50f363763c4639e2bf059335f656ddccf5f0c/boot/grubfm/init.sh#L76-L79
code: https://github.com/a1ive/grub/blob/50ac71615b4344fdf829c5c25f1f1e5680745eba/grub-core/grubfm/fm.c#L301

**Comment by a1ive** [comment]
> P.S. The files in grubfm \samples test for $grub_platform == "pc" - so they won't work if boot using elf iso?

I recommend
```
if [ "${grub_platform}" = "efi" ]; then
...
else
 # i386-pc and i386-multiboot
...
fi
```

**Comment by a1ive** [comment]
> it is possible to also UEFI32 and UEFI64 boot to EFI files too, even though we are in legacy mode (and the CPU will be in real mode)?

The grub under the multiboot platform acts like a real operating system. If we write a "uefi compatibility layer", then we can boot the efi file. And vice versa.

**Comment by steve6375** [comment]
OK. thanks very much. So basically all previous menuentries, scripts, etc. that test for grub_platform = pc will no longer work if using the elf version.
If using non-elf, then grub_mb_firmware is not set?
Why is is important for grub to set **grub_platform** to multiboot (instead of the standard strings of pc or efi) and break compatibility of all previous scripts and menus?  I don't see why you need to have a new value of 'multiboot' - why not just have a different variable if you need to know you have booted via elf?

**Reply by a1ive** [reply]
> So basically all previous menuentries, scripts, etc. that test for grub_platform = pc will no longer work if using the elf version.
> Why is is important for grub to set grub_platform to multiboot (instead of the standard strings of pc or efi) and break compatibility of all previous scripts and menus?

i386-multiboot is NOT i386-pc, they are 2 complete different platforms, and they load different modules.
This is because you (and some tutorials) have been assuming that grub_platform indicates the type of firmware, which is actually wrong.
Besides, I didn't invent a "multiboot" platform, it was always there in the official GNU GRUB version.

**Reply by a1ive** [reply]
I know many people misunderstand the meaning of this variable, so I still keep the ability to generate the i386-pc architecture grubfm.iso in the grubfm build script.

**Reply by a1ive** [reply]
https://github.com/a1ive/a1ive.github.io/blob/master/grub2_var_zh.md
> grub_platform
GRUB 构建时选择的平台类型， 如 "pc", "efi", "emu" 等。

google translation:
> grub_platform
The platform type selected during GRUB build, such as "pc", "efi", "emu", etc.

**Comment by steve6375** [comment]
> This is because you (and some tutorials) have been assuming that grub_platform indicates the type of firmware, which is actually wrong.

Not just me?  ;-)
![image](https://user-images.githubusercontent.com/1469350/106274984-0d3bb800-622d-11eb-9a61-317db602f95a.png)
